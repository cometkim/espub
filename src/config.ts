import * as fs from 'node:fs/promises';

export type Config = {
  name?: string,

  type?: 'commonjs' | 'module',

  /**
   * Source file for the `main`, `module`, and `exports` entry
   */
  source?: string,

  // Non-standard entry style for legacy bundlers
  module?: string,

  // Main entry
  main?: string,

  // Binary entries
  bin?: string | {
    [name: string]: string,
  },

  // TypeScript declaration for "main" entry
  types?: string,

  // Export maps
  exports?: ConditionalExport,

  dependencies?: {
    [name: string]: string,
  },

  peerDependencies?: {
    [name: string]: string,
  },

  browserslist?: string | string[],

  engines?: {
    node?: string,
  },
};

// See https://nodejs.org/api/packages.html#packages_nested_conditions
// What a mess :/
export type ConditionalExport = (
  | string
  | {
    [module: string]: ConditionalExport,
  }
  | {
    'import'?: ConditionalExport,
    'require'?: ConditionalExport,
    'node'?: ConditionalExport,
    'node-addons'?: ConditionalExport,
    'default'?: ConditionalExport,

    // community conditions definitions
    // See https://nodejs.org/api/packages.html#packages_community_conditions_definitions
    // See also https://devblogs.microsoft.com/typescript/announcing-typescript-4-7/#package-json-exports-imports-and-self-referencing
    'types'?: ConditionalExport,
    'deno'?: ConditionalExport,
    'browser'?: ConditionalExport,
    'development'?: ConditionalExport,
    'production'?: ConditionalExport,
  }
);

type ConfigWithOverride = Config & {
  publishConfig?: Config,
};

type LoadConfigOptions = {
  resolvePath: (path: string) => string,
};

export async function loadConfig({ resolvePath }: LoadConfigOptions): Promise<Config> {
  const configPath = resolvePath('package.json');

  const { publishConfig, ...config } = await fs.readFile(configPath, 'utf-8')
    .then(JSON.parse) as ConfigWithOverride;

  return {
    ...config,
    ...publishConfig,
  };
}
