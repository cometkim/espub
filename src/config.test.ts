import { resolve } from 'node:path';
import { describe, test, expect } from 'vitest';
import type { CompilerOptions as TSCompilerOptions } from 'typescript';

import { parseConfig } from './config';
import type { ParsedConfig } from './config';
import type { Flags } from './cli';
import type { Manifest } from './manifest';

describe('parseConfig', () => {
  const defaultFlags: Flags = {
    cwd: '/project',
    platform: undefined,
    rootDir: undefined,
    outDir: undefined,
    tsconfig: 'tsconfig.json',
    importMaps: 'package.json',
    external: [],
    standalone: false,
    noDts: false,
    noSourcemap: false,
  };

  const defaultManifest: Manifest = {
    name: 'package'
  };

  const defaultTsCompilerOptions: TSCompilerOptions = {
    target: 99, // ESNext,
    declaration: true,
  };

  test('validate manifest', () => {
    const result = parseConfig({
      flags: defaultFlags,
      manifest: defaultManifest,
    });

    expect(result).toEqual<ParsedConfig>({
      cwd: '/project',
      module: 'commonjs',
      platform: 'netural',
      sourcemap: true,
      declaration: false,
      rootDir: 'src',
      outDir: 'lib',
      manifest: defaultManifest,
    });
  });

  describe('flags', () => {
    test('--rootDir', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          rootDir: '.',
        },
        manifest: defaultManifest,
      });

      expect(result).toEqual<ParsedConfig>({
        cwd: '/project',
        module: 'commonjs',
        platform: 'netural',
        sourcemap: true,
        declaration: false,
        rootDir: '.',
        outDir: 'lib',
        manifest: defaultManifest,
      });
    });

    test('--outDir', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          outDir: '.',
        },
        manifest: defaultManifest,
      });

      expect(result).toEqual<ParsedConfig>({
        cwd: '/project',
        module: 'commonjs',
        platform: 'netural',
        sourcemap: true,
        declaration: false,
        rootDir: 'src',
        outDir: '.',
        manifest: defaultManifest,
      });
    });

    test('rootDir=outDir', () => {
      expect(() => parseConfig({
        flags: {
          ...defaultFlags,
          rootDir: '.',
          outDir: '.',
        },
        manifest: defaultManifest,
      })).toThrowErrorMatchingSnapshot();
    });
  });

  describe('tsCompilerOptions', () => {
    test('declaration=true', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        tsCompilerOptions: defaultTsCompilerOptions,
      });

      expect(result).toEqual<ParsedConfig>({
        cwd: '/project',
        module: 'commonjs',
        platform: 'netural',
        sourcemap: true,
        declaration: true,
        rootDir: 'src',
        outDir: 'lib',
        manifest: defaultManifest,
      });
    });

    test('declaration=false', () => {
      const result = parseConfig({
        flags: defaultFlags,
        manifest: defaultManifest,
        tsCompilerOptions: {
          ...defaultTsCompilerOptions,
          declaration: false,
        },
      });

      expect(result).toEqual<ParsedConfig>({
        cwd: '/project',
        module: 'commonjs',
        platform: 'netural',
        sourcemap: true,
        declaration: false,
        rootDir: 'src',
        outDir: 'lib',
        manifest: defaultManifest,
      });
    });

    test('declaration=true, --no-dts', () => {
      const result = parseConfig({
        flags: {
          ...defaultFlags,
          noDts: true,
        },
        manifest: defaultManifest,
        tsCompilerOptions: defaultTsCompilerOptions,
      });

      expect(result).toEqual<ParsedConfig>({
        cwd: '/project',
        module: 'commonjs',
        platform: 'netural',
        sourcemap: true,
        declaration: false,
        rootDir: 'src',
        outDir: 'lib',
        manifest: defaultManifest,
      });
    });
  });
});
