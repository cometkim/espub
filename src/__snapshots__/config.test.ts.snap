// Vitest Snapshot v1

exports[`parseConfig > flags > rootDir=outDir 1`] = `"Directory rootDir(.) and outDir(.) are conflict! Please specify different directory for one of them."`;
